### PR TITLE
FIX-#3292: Fix `shift` function when index has duplicates

### LINKED ISSUE #3292: Dataframe shift breaks with Modin distribution but not stock pandas
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Modin version** (`modin.__version__`): 0.10.1
- **Python version**: 3.7.9
- **Code we can use to reproduce**:
-  contact alyssa.unell@intel.com for information regarding the code and dataset

### Describe the problem
Shifting of a dataframe using modin distribution breaks

### Source code / logs
UserWarning: Distributing <class 'NoneType'> object. This may take some time.
Traceback (most recent call last):
  File "deeplearning/RecSys2020/02_ModelsCompetition/XGBoost2/sub040_1_save_public.py", line 731, in <module>
    add_timeshift(train, test, shift=1)
  File "deeplearning/RecSys2020/02_ModelsCompetition/XGBoost2/sub040_1_save_public.py", line 339, in add_timeshift
    gf["b_user_id_shifted"] = gf["b_user_id"].shift(periods=shift)
  File "/localdisk/alyssa/anaconda3/envs/aikit-modin/lib/python3.7/site-packages/modin/pandas/series.py", line 1361, in shift
    periods=periods, freq=freq, axis=axis, fill_value=fill_value
  File "/localdisk/alyssa/anaconda3/envs/aikit-modin/lib/python3.7/site-packages/modin/pandas/base.py", line 2319, in shift
    new_frame.index = self.index.copy()
  File "/localdisk/alyssa/anaconda3/envs/aikit-modin/lib/python3.7/site-packages/modin/pandas/base.py", line 562, in _set_index
    self._query_compiler.index = new_index
  File "/localdisk/alyssa/anaconda3/envs/aikit-modin/lib/python3.7/site-packages/modin/backends/pandas/query_compiler.py", line 91, in set_axis
    self._modin_frame.index = idx
  File "/localdisk/alyssa/anaconda3/envs/aikit-modin/lib/python3.7/site-packages/modin/engines/base/frame/data.py", line 250, in _set_index
    new_index = self._validate_set_axis(new_index, self._index_cache)
  File "/localdisk/alyssa/anaconda3/envs/aikit-modin/lib/python3.7/site-packages/modin/engines/base/frame/data.py", line 212, in _validate_set_axis
    "new values have %d elements" % (old_len, new_len)
ValueError: Length mismatch: Expected axis has 199999 elements, new values have 200000 elements

### PR DESCRIPTION
**Pull request by prutskov**
Signed-off-by: Alexey Prutskov <alexey.prutskov@intel.com>

## What do these changes do?

This PR does the next:

- fixes #3292 
- improves performance of `shift` operation

<details><summary>Performance results</summary>
The average time from 10 runs is shown (in seconds) in the tables below.

`df.shift(periods=10)`.
| nrows x ncols  | master  | PR #3345  | pandas  |
|---|---|---|---|
| 1k x 100 | 0.036  | 0.030| 0.0006  | 
| 1m x 100 |  0.871  | 0.042  | 0.729  | 
| 10m x 100  | 9.706   | 0.087 | 7.588  |
| 1k x 1k | 0.120   | 0.117 | 0.006   | 
| 10k x 10k | 1.062  | 1.087 | 0.733  | 
| 1m x 1k  | 1.124  |  0.293 | 7.156  | 

`df.shift(periods=500)`.
| nrows x ncols  | master  | PR #3345  | pandas  |
|---|---|---|---|
| 1k x 1k | 0.659   | 0.661 | 0.005 | 
| 10k x 10k | 3.498 | 3.480 | 0.710 | 
| 1m x 1k  | 1.626 |  0.833  | 7.114| 

<details><summary>Script</summary>

```python
import argparse
import time

import ray
import numpy as np
import modin.pandas as pd
import pandas as pd
ray.init()


def get_time(df, shift):
    start = time.time()
    df.shift(periods=shift)
    duration = time.time() - start
    return duration

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Benchmark shift")
    parser.add_argument("--num_repeats", type=int, default=5)
    parser.add_argument("--num_rows", type=int, default=1_000)
    parser.add_argument("--num_cols", type=int, default=100)
    parser.add_argument("--shift", type=int, default=10)

    args = parser.parse_args()

    data = np.random.randint(0, 100, size=(args.num_rows, args.num_cols))
    df = pd.DataFrame(data)

    times = []
    for i in range(args.num_repeats):
        times.append(get_time(df, args.shift))
        print(times[-1])

    print(
    f"1st iter: {times[0]:5f} s, min: {min(times):5f} s, aver: {sum(times) / args.num_repeats:5f} s"
)

```

</details>

The results in synchronous mode (request from https://github.com/modin-project/modin/pull/3345#issuecomment-901359611):

`execute(df.shift(periods=10))`.
| nrows x ncols  | master  | PR #3345  | pandas  |
|---|---|---|---|
| 1k x 100 | 0.125 | 0.136| 0.0006  | 
| 1m x 100 |  1.202  | 0.377  | 0.729  | 
| 10m x 100  | 10.560   | 1.012 | 7.588  |
| 1k x 1k | 0.635   | 0.627 | 0.006   | 
| 10k x 10k | 6.953  | 6.985 | 0.733  | 
| 1m x 1k  | 3.613  |  2.667 | 7.156  | 

`execute(df.shift(periods=500))`.
| nrows x ncols  | master  | PR #3345  | pandas  |
|---|---|---|---|
| 1k x 1k | 1.165   | 1.132 | 0.005 | 
| 10k x 10k | 10.199 | 10.252 | 0.710 | 
| 1m x 1k  | 4.417 |  3.329  | 7.114| 

</details>

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3292 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by anmyachev** [comment]
@prutskov to take into account the effect of lazy evaluation, we also need to call the function https://github.com/modin-project/modin/blob/f74eec3b2c807d13721d367f62ee3ba18dcd0ba7/asv_bench/benchmarks/utils.py#L490

```python
execute(df.shift(...))
```

What performance numbers do you get in the case?

**Comment by prutskov** [comment]
@anmyachev The results in synchronous mode were added in performance section of [PR description](https://github.com/modin-project/modin/pull/3345#issue-714174480).

**Comment by anmyachev** [comment]
> @anmyachev The results in synchronous mode were added in performance section of [PR description](https://github.com/modin-project/modin/pull/3345#issue-714174480).

Thanks @prutskov!

File: modin/pandas/base.py [hunk 1]
```diff
@@ -2336,20 +2336,15 @@ def shift(self, periods=1, freq=None, axis=0, fill_value=no_default):
...
-                    new_frame = dropped_df.append(filled_df, ignore_index=True)
-                    new_frame.index = self.index.copy()
-                    if isinstance(self, DataFrame):
-                        new_frame.columns = self.columns.copy()
-                    return new_frame
+                new_frame = (
+                    filled_df.append(self.iloc[:-periods], ignore_index=True)
+                    if periods > 0
+                    else self.iloc[-periods:].append(filled_df, ignore_index=True)
+                )
```
Comment (anmyachev) [inline modin/pandas/base.py:2343]: It turns out that the speedup is achieved by using `iloc` instead of `drop`?

Comment (prutskov) [reply modin/pandas/base.py:2343]: Yes, only `drop`->`iloc` for case `shift(axis=0)`. Moreover, I tried include this change in branch for `axis=1`, but got worse performance results.

**Review by anmyachev — APPROVED** [review]

**Review by YarShev — APPROVED** [review]

**Pull request merged**